feat: フォローボタンのデザインをわかりやすくする

sanao1006 opened the pull request:
# 概要
以前までは、プロフィール画面、ユーザ一覧画面のフォローボタンのデザインがフォロー・非フォロー状態に関わらず同じような見た目で分かりづらかった。 

# 変更点
- プロフィール画面のフォローボタンのデザインを、ユーザー一覧画面のようにフォロー・非フォロー状態次第で見た目を切り替える

## スクリーンショット(任意)
フォロー中 | 非フォロー |  非フォロー申請済み | 非フォロー未申請 |
:--: | :--:| :--:| :--:
![image](https://github.com/pantasystem/Milktea/assets/62137820/15950ce3-01cb-4758-81a3-dcc5fc56fa62)|![image](https://github.com/pantasystem/Milktea/assets/62137820/8875ce9f-ecd0-4665-a8c4-faa01a0268e9)|![image](https://github.com/pantasystem/Milktea/assets/62137820/7c10743d-e544-4b20-817a-9d9bc9a6f88a)|![image](https://github.com/pantasystem/Milktea/assets/62137820/ea36022d-dd28-4d62-939a-612f9ceaa025)|


## Issue番号
Fixes #1955

sanao1006 committed:
feat: フォローボタンのデザインをわかりやすくする

以前までは、プロフィール画面、ユーザ一覧画面のフォローボタンのデザインがフォロー・非フォロー状態に関わらず同じような見た目で分かりづらかった。
よって、プロフィール画面のフォローボタンのデザインを、ユーザー一覧画面のようにフォロー・非フォロー状態次第で見た目を切り替えるようにする

Fixes #1955

sanao1006 commented on modules/features/user/src/main/java/net/pantasystem/milktea/user/profile/UserDetailActivity.kt:
```diff
@@ -472,6 +509,13 @@ class UserDetailActivity : AppCompatActivity() {
         binding.remoteUserState.setBackgroundColor(typed.data)
     }
 
+    private fun setFollowState(state: FollowState?): String = when (state) {
```
["followState"BindingAdapter](https://github.com/pantasystem/Milktea/blob/c934497ab4f34a12be23cbb097c02ef3be94c9ae/modules/features/user/src/main/java/net/pantasystem/milktea/user/helper/FollowStateHelper.kt#L12)をいい感じに変形したりして再利用したかったが、Jetpack ComposeとBindingAdapterの合わせ方がよくわからなかったためこの関数を作成した

pantasystem replied:
https://github.com/pantasystem/Milktea/pull/1961/files#diff-c2246cd8b802c5448fccf91488f790ff9f7d8884b207d7773a68d7eaf24f951dR203-R221 をComposable関数で切り出した上で、[stringResource](https://developer.android.com/jetpack/compose/resources?hl=ja)関数で取得するのが良いと思いました。

pantasystem commented on modules/features/user/src/main/java/net/pantasystem/milktea/user/profile/UserDetailActivity.kt:
```diff
@@ -187,6 +193,37 @@ class UserDetailActivity : AppCompatActivity() {
+                val isMine = mViewModel.isMine.collectAsState()
+                val buttonText = setFollowState(userDetail?.followState)
+                MdcTheme {
+                    if (!isMine.value) {
+                        when (userDetail?.followState) {
+                            FollowState.UNFOLLOWING, FollowState.UNFOLLOWING_LOCKED -> {
+                                OutlinedButton(
+                                    shape = RoundedCornerShape(32.dp),
+                                    onClick = { mViewModel.changeFollow() },
+                                ) {
+                                    Text(buttonText)
+                                }
+                            }
+                            FollowState.FOLLOWING, FollowState.PENDING_FOLLOW_REQUEST -> {
+                                Button(
+                                    shape = RoundedCornerShape(32.dp),
+                                    onClick = { mViewModel.changeFollow() },
+                                ) {
+                                    Text(buttonText)
+                                }
+                            }
+                            else -> {  }
+                        }
+                    }
+                }
+            }
```
Composable関数に切り出すて実装するとよりよく実装できるのではないかと思いました。

pantasystem reviewed (commented):
ご確認よろしくお願いします🙇

sanao1006 committed:
fix: フォローボタン実装を修正

- フォローボタンをComposable関数に切り出した
- FollowStateの文字列リソースをstringResourceで取得するように変更

Fixes #1955

sanao1006 commented:
修正しました！
確認よろしくおねがいします！

sanao1006 commented:
メモリエラーですか🤔

onCreateの中にComposable関数を書いたのがよくなかったかもしれない...？

sanao1006 committed:
fix: `FollowButton`をonCreateの外に出す

Fixes #1955

pantasystem commented on modules/features/user/src/main/java/net/pantasystem/milktea/user/profile/UserDetailActivity.kt:
```diff
@@ -187,6 +197,21 @@ class UserDetailActivity : AppCompatActivity() {
+                    FollowButton(
+                        userDetail = userDetail,
+                        isMine = isMine,
+                        buttonText = buttonText,
```
buttonTextの取得処理はFollowButtonの中に閉じてしまうのが好ましいと思いました。
理由としてはFollowButton自体が状態を受け取り、
それに応じて表示内容を切り替えているので、FollowButtonの中に閉じてしまうのが良いと思いました。

pantasystem commented on modules/features/user/src/main/java/net/pantasystem/milktea/user/profile/UserDetailActivity.kt:
```diff
@@ -472,6 +497,44 @@ class UserDetailActivity : AppCompatActivity() {
+    fun FollowButton(
+        userDetail: User.Detail?,
+        isMine: State<Boolean>,
+        buttonText: String,
```
userDetailは直値指定で、isMineがState経由なのが気になりました。
何かわけがあるのでしょうか？

sanao1006 replied:
すみません、これはおそらく
```kotlin
val isMine = mViewModel.isMine.collectAsState()
```
のところで、`by`を間違えて`=`にしていたからですね。
普通にミスっていました🙇

pantasystem commented on modules/features/user/src/main/java/net/pantasystem/milktea/user/profile/UserDetailActivity.kt:
```diff
@@ -472,6 +497,44 @@ class UserDetailActivity : AppCompatActivity() {
+                        Text(buttonText)
+                    }
+                }
+                else -> {  }
```
elseを使用してしまうと、将来的にFollowStateに状態が追加された時に、
その変更を見落としてしまうことに繋がりかねないので、
null -> {} にしたほうが良いと思いました。

pantasystem commented on modules/features/user/src/main/java/net/pantasystem/milktea/user/profile/UserDetailActivity.kt:
```diff
@@ -472,6 +497,44 @@ class UserDetailActivity : AppCompatActivity() {
+                FollowState.UNFOLLOWING, FollowState.UNFOLLOWING_LOCKED -> {
+                    OutlinedButton(
+                        shape = RoundedCornerShape(32.dp),
+                        onClick = { mViewModel.changeFollow() },
```
直接ViewModelを制御してしまうとComposable関数としての再利用性、テスト容易性が低下してしまうので、
一度コールバックとしてイベントを戻してあげて、そこでViewModelにイベントを伝えてあげるのが良いと思いました。
また別ファイルとしてComposable関数として切り出すことによって、
フォロー一覧画面などで使われている、[UserDetailCard](https://github.com/pantasystem/Milktea/blob/c934497ab4f34a12be23cbb097c02ef3be94c9ae/modules/features/user/src/main/java/net/pantasystem/milktea/user/compose/UserDetailCard.kt#L40-L40)でも再使用できるようになりそうだと思いました。

sanao1006 replied:
いいと思います！！

それとは別に思ったんですけど、[serStateActionButton](https://github.com/pantasystem/Milktea/blob/c934497ab4f34a12be23cbb097c02ef3be94c9ae/modules/features/user/src/main/java/net/pantasystem/milktea/user/compose/UserDetailCard.kt#L219)にならってFollowButtonには`FollowState?`を直接渡すのってどうですかね？

userDetailはfollowStateしか使ってないのでこの方が軽くなるかなと思ったからです

pantasystem replied:
良いと思います
あとはPreviewを表示する時に渡す引数を少なくできて良さそうだと思いました

pantasystem reviewed (commented):
ご確認お願いします。

sanao1006 committed:
fix: `FollowButton`の実装を修正

- buttonTextの取得処理はFollowButton自体が状態を受け取りそれに応じて表示内容を切り替えているため、FollowButtonの中に閉じ込める
- isMineをBool直値指定にする
- elseをnullに変更する
- FollowButtonの再利用性を上げるため、クリック時のイベントにはコールバック関数を使用する
- テスト容易性が良くないので、FollowButtonには`FollowState?`を直接渡す

Fixes #1955

sanao1006 commented:
@pantasystem 

お手すきの際にご確認お願いします🙇

pantasystem commented on modules/features/user/src/main/java/net/pantasystem/milktea/user/profile/UserDetailActivity.kt:
```diff
@@ -472,6 +495,45 @@ class UserDetailActivity : AppCompatActivity() {
     }
 
+    @Composable
+    fun FollowButton(
```
別ファイルに切り出したほうが視認性が良くなると思いました。
当初は文字列リソースを得るためにcontextへの参照が必要なため
Activity中に実装していたと思うのですが、
stringResourceは直接的なContextへの参照は不要なため、別ファイルに切り出せると思います。

sanao1006 replied:
FollowButtonを切り出したファイルの置き場所ですが、`modules/features/user/.../user/compose`の中で問題ないでしょうか？
それとも`modules/features/user/.../user`直下あたりでしょうか。微妙に迷っています。。。

pantasystem replied:
今回はこの機能（画面）で使われることを前提としているので、`modules/features/user/.../user` で良いかと思われます。

pantasystem commented on modules/features/user/src/main/java/net/pantasystem/milktea/user/profile/UserDetailActivity.kt:
```diff
@@ -472,6 +495,45 @@ class UserDetailActivity : AppCompatActivity() {
+    }
+
+    @Composable
+    private fun setFollowState(state: FollowState?): String = when (state) {
```
実際には値をsetしていないので、getXXXなどにリネームしていただきたいです。

pantasystem reviewed (changes requested):
お手隙の際にご確認よろしくお願いします。

sanao1006 committed:
refactor: FollowButtonをファイルに切り出す

- UserDetailActivityの内部にあったFollowButton関数をファイルに切り出した
- setFollowState関数をリネームした

pantasystem reviewed (approved):
お疲れ様でした！
コントリビュートありがとうございました🙇
![LGTM](https://image.lgtmoon.dev/72640)

sanao1006 commented:
こちらこそ丁寧にレビューありがとうございました‼️
